### ISSUE TITLE
on behaviour of `OutputModules` in unscheduled execution, and its interplay with HLT

### ISSUE THREAD

**Issue by missirol** [cms-sw/cmssw#36138]
As suggested by @silviodonato, I'm opening an issue related to a recent discussion between Core-Sw and HLT concerning the behaviour of `OutputModules` in unscheduled execution, to keep better track of this problem and its possible solutions (the original exchange was had in the Core-Sw Mattermost channel on Nov, 8-9).

The use case is HLT's goal to fully support workflows with `Tasks` and `SwitchProducers` where needed in the HLT menu for Run-3 (e.g. triggers with GPU offloading).

I'm copying below part of the original question [*].

FYI: @fwyzard @sharper @Martin-Grunewald @makortel @Dr15Jones

[*]

> _Context_ : in Run-2 HLT did not make use of `Tasks`; in Run-3 it will, mainly b/c of the use of `SwitchProducers` in triggers with GPU offloading. We are currently testing if it's possible to move to `Tasks` in other parts of the HLT menu, beyond what would be strictly necessary. Note that "migrating HLT to Tasks" for us currently means that `EDProducers` of a given `Path` are moved to a `Task`, but `EDFilters` are not (they remain directly on the Path, or on a Sequence used in the Path), because the filter decision is needed (as known, in a HLT path one wants to reject the event as soon as possible, without running producers unnecessarily).
>
>  After migrating all paths of an HLT config to Tasks as a test, we encountered problems related to `OutputModules`.
>
> I tried to exemplify the issue in [this dummy config](https://gist.github.com/missirol/29e44a3b2d6c79894674950be9ac245c). There is 1 `OutputModule` writing events for the OR of 2 triggers. The `OutputModule` expects to write the products of 1 producer which runs unscheduled. This producer (`theVerticesAfterStep2`) is associated to 1 Path (`thePathWithProducts`) and, in turn, it needs the products of an EDFilter in the same Path, but said EDFilter (`theVerticesAfterStep1`) runs scheduled. The other trigger (`thePathAlwaysTrue`) always fires, while `thePathWithProducts` fails before the filter `theVerticesAfterStep1` is executed; the `OutputModule` tries to run `theVerticesAfterStep2` (unscheduled), and this leads to a crash b/c the products of `theVerticesAfterStep1` are missing.
>
> The issue boils down to the fact that, as explained in the [documentation](https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideUnscheduledExecution#Usage), `OutputModules` can trigger the unscheduled execution of a module if its products are expected to be saved in the output file.
>
> Another way this could complicate things for HLT is shown in a [2nd example](https://gist.github.com/missirol/93743b27bcabe01c1fec0c74cbd35c96). Here again, we have an `OutputModule` writing events for the OR of 2 triggers. A common situation for HLT is that one path gets disabled (or, 'prescaled to zero') to avoid running certain modules (here, `theVerySlowModule`). But, since `theVerySlowModule` is unscheduled and expected by the `OutputModule`, it is executed despite `thePathWithSlowModules` being "disabled". Clearly, the Fwk is working as intended, just not in the most convenient/intuitive way for HLT.
>
> Question is: would it make sense to have the option to disable the ability of `OutputModules` to require the unscheduled execution of modules (in other words, simply having `outputCommands` as a list of directives for saving products *only if* they were produced)? If yes, how difficult would that be to implement?

**Comment by Martin-Grunewald** [comment]
assign hlt

**Comment by Martin-Grunewald** [comment]
assign core

**Comment by makortel** [comment]
We're planning to introduce a new Path type, `cms.FinalPath`. Only OutputModules would be allowed to be inserted in those, and those OutputModules would be run for an Event only after all other EDModules have been run for that Event, and those OutputModules would not trigger prefetching of the consumed data products (i.e. those OutputModules alone would not cause any other EDModule to run).

**Comment by Martin-Grunewald** [comment]
Hmm, I do not know whether this restriction (only OMs on FinalPaths) would work: within the HLT menu, we currently have OMs on EndPaths which also have Filters (eg, HLTPrescaler and/or TriggerResultsFilter) and Producers (eg, Scouting) on these EndPaths (in front of their OM).

**Comment by makortel** [comment]
Thanks @Martin-Grunewald. Naively the current EDFilters in an EndPath could be moved to another regular Path, and configure the OutputModule to store only those Events passing that Path (via `SelectEvents`). But could you remind us, again, on the use cases for (specific) EDFilters in EndPaths?

**Comment by fwyzard** [comment]
Hi Matti,

the most common use case is to apply prescales and "smart prescales" (i.e.
per-trigger-path prescales and other logical expressions) in addition to
the event selection applied by the OutputModule itself.
For example, the Express and DQM streams apply both an event selection in
the OutputModule and a more dynamic preselaction via an EDFilter in the
EndPath.

An other use case (that may be implemented using dedicated prescales, or a
different filtrt altogether) is to enable or disable a DAQ stream (i.e. an
output file, so an OutputModule) dynamically during the run.

I agree that it should be possible to implement these use cases with an
additional dedicated Path; in fact, we may do it anyway to simplify the
definition of the DAQ streams and primary datasets.

There may be an additional use case: splitting a DAQ stream and its
datasets into N copies; I'm not sure if we can cover this use case without
EDFilters on the FinalPaths.

.A

**Comment by makortel** [comment]
Thanks Andrea.

In principle a sequence of EDFilters is equivalent to logical AND between the EDFilter return values. We could enhance the `SelectEvents` ability of OutputModule to evaluate an AND of given Paths, such that this AND would be evaluated before the OR of current `SelectEvents` (although strictly speaking the same effect can be achieved already now with specific structure of Paths and `PathStatusFilter`).

**Comment by Martin-Grunewald** [comment]
Well, the HLTPrescaler and TriggerResultsFilter can not be simulated by an AND of the paths, as individual path-dependent or global prescales are applied. Do you suggest that for each stream/OM, we first define a path which does that prescaling, and then the endpath with the OM then refers only to that path? That seems cumbersome and error prone.

**Comment by fwyzard** [comment]
I think the idea of reorganising the streams and datasets we discussed with the Heavy Ions group already goes in that direction.

Here is an example of how we could organise a simple HLT menu following that approach.
The outcome should be a menu with
  - **PhysicsCommissioning** stream
      - _Commissioning_part1_ dataset
      - _Commissioning_part2_ dataset
  - **PhysicsEGamma** stream
      - _SingleElectron_ dataset
  - **PhysicsMuons** stream
      - _SingleMuon_ dataset
      - _DoubleMuon_ dataset
  - **Express** stream
      - _Express_ dataset

Note how with this approach we split the _Commissioning_part1_ and _Commissioning_part2_ dataset without the need to duplicate all the paths they contain.

```python
# zero bias and commissioning paths
HLT_ZeroBias = cms.Path(...)
...

# electron trigger paths
HLT_SingleElectron_Loose = cms.Path(...)
HLT_SingleElectron_Tight = cms.Path(...)
...

# muon trigger paths
HLT_SingleMuon = cms.Path(...)
HLT_IsoSingleMuon = cms.Path(...)
HLT_DoubleMuon = cms.Path(...)
...

# definition of the primary datasets as the logical OR of various trigger paths
#   - the HLTPrescaler can be used to split a dataset (and all the paths it contains) into multiple slices
#   - the TriggerResultsFilters can add any smart prescales, e.g. for DQM or Express

datasetCommissioningPart1Prescaler = cms.EDFilter("HLTPrescaler",
    offset = cms.uint32(0),
    ...
)
datasetCommissioningPart2Prescaler = cms.EDFilter("HLTPrescaler",
    offset = cms.uint32(1),
    ...
)
...
datasetCommissioningDefinition = cms.EDFilter("TriggerResultsFilter",
    usePathStatus = cms.bool(True),
    hltResults = cms.InputTag(""),
    l1tResults = cms.InputTag(""),
    triggerConditions = cms.vstring("HLT_ZeroBias", ...)
)
dataset_Commissioning_part1 = cms.Path(datasetCommissioningPart1Prescaler + datasetCommissioningDefinition)
dataset_Commissioning_part2 = cms.Path(datasetCommissioningPart2Prescaler + datasetCommissioningDefinition)
...

datasetSingleElectronPrescaler = cms.EDFilter("HLTPrescaler", ...)
datasetSingleElectronDefinition = cms.EDFilter("TriggerResultsFilter",
    usePathStatus = cms.bool(True),
    hltResults = cms.InputTag(""),
    l1tResults = cms.InputTag(""),
    triggerConditions = cms.vstring("HLT_SingleElectron_Loose", "HLT_SingleElectron_Tight", ...)
)
dataset_SingleElectron = cms.Path(datasetSingleElectronPrescaler + datasetSingleElectronDefinition)

datasetSingleMuonPrescaler = cms.EDFilter("HLTPrescaler", ...)
datasetSingleMuonDefinition = cms.EDFilter("TriggerResultsFilter",
    usePathStatus = cms.bool(True),
    hltResults = cms.InputTag(""),
    l1tResults = cms.InputTag(""),
    triggerConditions = cms.vstring("HLT_SingleMuon", "HLT_IsoSingleMuon", ...)
)
dataset_SingleMuon = cms.Path(datasetSingleMuonPrescaler + datasetSingleMuonDefinition)

datasetDoubleMuonPrescaler = cms.EDFilter("HLTPrescaler", ...)
datasetDoubleMuonDefinition = cms.EDFilter("TriggerResultsFilter",
    usePathStatus = cms.bool(True),
    hltResults = cms.InputTag(""),
    l1tResults = cms.InputTag(""),
    triggerConditions = cms.vstring("HLT_DoubleMuon", ...)
)
dataset_DoubleMuon = cms.Path(datasetDoubleMuonPrescaler + datasetDoubleMuonDefinition)

datasetExpressPrescaler = cms.EDFilter("HLTPrescaler", ...)
datasetExpressDefinition = cms.EDFilter("TriggerResultsFilter",
    usePathStatus = cms.bool(True),
    hltResults = cms.InputTag(""),
    l1tResults = cms.InputTag(""),
    triggerConditions = cms.vstring("HLT_ZeroBias / 20", "HLT_SingleElectron_Loose / 10", "HLT_SingleMuon / 10", "HLT_IsoSingleMuon / 5", ...)
)
dataset_Express = cms.Path(datasetExpressPrescaler + datasetExpressDefinition)


# definition of the DAQ streams as the logical OR of various primary datasets;
# additional filters or producers should not be necessary here

streamPhysicsCommissioningOutput = cms.OutputModule( "PoolOutputModule",
    ...,
    SelectEvents = cms.untracked.PSet( SelectEvents = cms.vstring(
        "dataset_Commissioning_part1",
        "dataset_Commissioning_part2",
        ...
    ))
)
stream_PhysicsCommissioning = cms.FinalPath(streamPhysicsCommissioningOutput)

streamPhysicsEGammaOutput = cms.OutputModule( "PoolOutputModule",
    ...,
    SelectEvents = cms.untracked.PSet( SelectEvents = cms.vstring(
        "dataset_SingleElectron",
        ...
    ))
)
stream_PhysicsEGamma = cms.FinalPath(streamPhysicsEGammaOutput)

streamPhysicsMuonsOutput = cms.OutputModule( "PoolOutputModule",
    ...,
    SelectEvents = cms.untracked.PSet( SelectEvents = cms.vstring(
        "dataset_SingleMuon",
        "dataset_DoubleMuon",
        ...
    ))
)
stream_PhysicsMuons = cms.FinalPath(streamPhysicsMuonsOutput)

streamExpressOutput = cms.OutputModule( "PoolOutputModule",
    ...,
    SelectEvents = cms.untracked.PSet( SelectEvents = cms.vstring(
        "dataset_Express",
        ...
    ))
)
stream_Express = cms.FinalPath(streamExpressOutput)
```

(names, prescales, _etc._ are just a made up example; the actual naming convention, whether we number datasets from 0 or 1, whether to have prescales in the physics datasets, _etc._ are still to be defined)

If I didn't miss anything, we should not need any EDProducers or EDFilters in the FinalPaths.

**Comment by fwyzard** [comment]
@silviodonato FYI

**Comment by silviodonato** [comment]
I think that the FinalPath + Andrea's implementation should be ok from the trigger point of view.
About prescales:
- nothing would change in prescale of Path
- we would be able to apply prescales and smart prescales to Dataset instead of Stream. And this is a good news as it will increase our flexibility

**Comment by fwyzard** [comment]
In case we need to add prescales to split _streams_ instead of datasets (which would probably break the current Tier-0 processing, but hey, we can try...) we can always add an other level of `Paths` for the streams like we will do for the datasets.

**Comment by makortel** [comment]
@fwyzard A question mostly out of curiosity, in 
```py
stream_PhysicsCommissioning = cms.FinalPath(streamPhysicsCommissioningOutput)
...
stream_PhysicsEGamma = cms.FinalPath(streamPhysicsEGammaOutput)
...
stream_Express = cms.FinalPath(streamExpressOutput)
```
do you see benefits in having a separate `FinalPath` for each OutputModule instead of putting all of them into a one `FinalPath` along
```py
outputModules = cms.FinalPath(
    streamPhysicsCommissioningOutput,
    streamPhysicsEGammaOutput,
    streamExpressOutput
)
```
?

**Comment by fwyzard** [comment]
One of the reasons for multiple EndPaths was to have a different EDFilter
on each one, to enable or disable the corresponding DAQ stream.

With the new approach, this could be done at the dataset level, or at the
stream level adding an extra layer of paths.

So, if the FinalPaths cannot have EDFilters, there is probably is no need
for having multiple of them - other than the convenience of adding/removing
them to/from a menu.

**Comment by Martin-Grunewald** [comment]
Well, while this is all technically possible, I am still of the opinion that this makes the config way more complicated and error prone as now several paths are involved, as so far this has all to be configured and managed manually.

**Comment by Sam-Harper** [comment]
@Martin-Grunewald 

I think this involves having some logic and checks in confdb to handle this.  So far I dont see a major show stopper.  And we want to revisit the datasets to something like this anyway

The only alterative solution I can see which preserves the HLT exactly as is  is  to let modules have a way of registering themselves outside unscheduled execution and run that group of modules last (and only run them if the execution makes it along the path to them). However it may be that the scheduler can not do this without a lot of work. And it also has the disadvantage that  its surprising to the user (ie theres no clear way to know a module does it without looking at the code).  Cleaner to stick it on a cms.FinalPath in my opinion. 

We will have to think about the confdb solution, while cms.FinalPath is fine (paths have status code which can range from 0-9 with only 0&1 used to tell what type of path it is), the logic to make things means we might not be able to make configs in older releases any more if we dont have some way of switching it off (which is likely more work).

**Comment by fwyzard** [comment]
> I am still of the opinion that this makes the config way more complicated and error prone as now several paths are involved,

I don't think so: the latest GRun menu has 58 paths that are duplicated of each other in order to be split across datasets; with the new approach they will no longer be needed.


> as so far this has all to be configured and managed manually.

As far as I know @Sam-Harper is working to make this possible in ConfDB.

**Comment by Martin-Grunewald** [comment]
Yes indeed, it simplifies that monstrosity, though makes other areas more complicated. Hmm, "working to make this possible in ConfDB" has an undefined timescale, so until it is realised we are stuck with some /different/ complication.

**Comment by Sam-Harper** [comment]
it is an undefined timescale of course. But it can be bumped up priority if there is a pressing need. And this feels like a pressing need.

**Comment by makortel** [comment]
Thanks @silviodonato. The last open prerelease, 12_2_0_pre3, next week is unfortunately not realistic. A backport before the final 12_2_0 release might be possible (I can comment more on the feasibility next week).

**Comment by silviodonato** [comment]
Yes, it is sufficient. As you said, the plan is to move the `EDFilters` that are currently in the `EndPath` to the new "dataset" `Paths`.
Each `FinalPath` should contain an `PoolOutputModule` and the `SelectEvents` should contain the OR of the new "dataset" `Paths`. (See the example in https://github.com/cms-sw/cmssw/issues/36138#issuecomment-972717360)

**Comment by makortel** [comment]
The `FinalPath` is implemented in #36398, and backport to 12_2_X should be feasible.

**Comment by makortel** [comment]
Backport to 12_2_X was done in https://github.com/cms-sw/cmssw/pull/36438 that is already merged.

**Comment by makortel** [comment]
+core

**Comment by missirol** [comment]
Back to this old issue, to try and summarise the outcome from the HLT side.

 - This issue led to the introduction of `cms.FinalPath`. This was ultimately not considered sufficient to support the Run-3 HLT menu, and was complemented by the introduction of `cms.ConditionalTask` (see [#36938](https://github.com/cms-sw/cmssw/issues/36938)).
 - `cms.FinalPath` was used in the Run-3 HLT menus of 2022, 2023 and 2024.
 - `cms.ConditinalTask` was used in the Run-3 HLT menus of 2022, 2023 and the first part of 2024.
 - Once the migration of the Run-3 HLT configurations to Alpaka-based modules was completed in 2024, both `cms.FinalPath` and `cms.ConditionalTask` became unnecessary for HLT's purposes.
 - [*] provides some more details on the use of `cms.FinalPath` and `cms.(Conditional)Task` in the Run-3 HLT (offline) menus maintained by TSG.

Many thanks to the Core-sw group for implementing these updates to HLT's benefit.

From my point of view, this issue can be closed.

[*]

 - Oct 2021: [cms-sw/hlt-confdb#26](https://github.com/cms-sw/hlt-confdb/pull/26) added support for `cms.Task` in the ConfDB GUI.

 - Feb 2022: [cms-sw/hlt-confdb#40](https://github.com/cms-sw/hlt-confdb/pull/40) added support for `cms.FinalPath` in the ConfDB GUI.

 - Mar 2022: [CMSHLT-2245](https://its.cern.ch/jira/browse/CMSHLT-2245) introduced `cms.FinalPath`s in the Run-3 HLT menus. This change was propagated to the HLT menus in CMSSW via [#37339](https://github.com/cms-sw/cmssw/pull/37339) (`CMSSW_12_3_0`) and [#37338](https://github.com/cms-sw/cmssw/pull/37338) (`CMSSW_12_4_0_pre2`).

 - May 2022: [CMSHLT-2288](https://its.cern.ch/jira/browse/CMSHLT-2288) introduced `cms.Task`s in the Run-3 HLT menus. This change was propagated to the HLT menus in CMSSW via [#37793](https://github.com/cms-sw/cmssw/pull/37793) (`CMSSW_12_3_3`) and [#37783](https://github.com/cms-sw/cmssw/pull/37783) (`CMSSW_12_4_0_pre4`).

 - Jun 2022: [cms-sw/hlt-confdb#49](https://github.com/cms-sw/hlt-confdb/pull/49) updated ConfDB to make use of `cms.ConditionalTask` (instead of `cms.Task`) for any HLT menu based on `CMSSW_12_4_0` and higher (except for `CMSSW_12_5_0_pre1`). This change was propagated to the HLT menus in CMSSW via [#38483](https://github.com/cms-sw/cmssw/pull/38483) (`CMSSW_12_4_1`) and [#38482](https://github.com/cms-sw/cmssw/pull/38482) (`CMSSW_12_5_0_pre3`).

 - Nov 2023: a plan to remove `cms.FinalPath`, `cms.ConditionalTask` and `SwitchProducer(CUDA)` from the HLT configurations, and eventually from CMSSW itself, was discussed among TSG and Core-sw (see minutes of [this meeting](https://indico.cern.ch/event/1345902)).

 - Sep 2024: [CMSHLT-3284](https://its.cern.ch/jira/browse/CMSHLT-3284) completed the migration of the Run-3 HLT menus to Alpaka-based modules, removing the last instances of `cms.ConditionalTask` and `SwitchProducer(CUDA)` from said HLT configurations. This change was propagated to the HLT menus in CMSSW via [#46050](https://github.com/cms-sw/cmssw/pull/46050) (`CMSSW_14_0_17`), [#46049](https://github.com/cms-sw/cmssw/pull/46049) (`CMSSW_14_1_0`), and [#46048](https://github.com/cms-sw/cmssw/pull/46048) (`CMSSW_14_2_0_pre2`).

 - Mar 2025: [CMSHLT-3434](https://its.cern.ch/jira/browse/CMSHLT-3434) removed all the instances of `cms.FinalPath` in the Run-3 HLT menus, replacing them with `cms.EndPath`. This change was propagated to the HLT menus in CMSSW via [#47530](https://github.com/cms-sw/cmssw/pull/47530) (`CMSSW_15_0_2`) and [#47529](https://github.com/cms-sw/cmssw/pull/47529) (`CMSSW_15_1_0_pre1`).

**Comment by makortel** [comment]
@cmsbuild, please close

**Comment by makortel** [comment]
`FinalPath` was removed in https://github.com/cms-sw/cmssw/pull/47546
